llpc/patch/llpcPatchIntrinsicSimplify.cpp

```diff
@@ -0,0 +1,416 @@
+    analysisUsage.addPreserved<ScalarEvolutionWrapperPass>();
+    analysisUsage.addRequired<PipelineStateWrapper>();
+    analysisUsage.setPreservesCFG();
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM function.
+bool PatchIntrinsicSimplify::runOnFunction(
+    Function& func) // [in,out] LLVM function to be run on.
+{
+    bool changed = false;
+
+    m_pPipelineState = getAnalysis<PipelineStateWrapper>().GetPipelineState(func.getParent());
+
+    m_pScalarEvolution = &getAnalysis<ScalarEvolutionWrapperPass>().getSE();
+    m_pContext = &func.getContext();
+    m_pFunction = &func;
+
+    for (Function& otherFunc : func.getParent()->functions())
+    {
+        // Skip non intrinsics.
+        if (otherFunc.isIntrinsic() == false)
+        {
+            continue;
+        }
+
+        SmallVector<Value*, 4> usersOfOtherFunc(otherFunc.users());
+
+        for (Value* const pUser : usersOfOtherFunc)
+        {
+            IntrinsicInst* const pIntrinsicCall = dyn_cast<IntrinsicInst>(pUser);
+
+            // Skip non-intrinsics.
+            if (pIntrinsicCall == nullptr)
+            {
+                continue;
+            }
+
+            // Skip calls not from our own function.
+            if (pIntrinsicCall->getFunction() != &func)
+            {
+                continue;
+            }
+
+            Value* const pSimplifiedValue = Simplify(pIntrinsicCall);
+
+            // If the returned value is nullptr, then we did not simplify it.
+            if (pSimplifiedValue == nullptr)
+            {
+                continue;
+            }
+
+            changed = true;
+
+            pIntrinsicCall->replaceAllUsesWith(pSimplifiedValue);
+            pIntrinsicCall->eraseFromParent();
+            m_pScalarEvolution->eraseValueFromMap(pIntrinsicCall);
+        }
+    }
+
+    return changed;
+}
+
+// =====================================================================================================================
+// Check if a value is safely derived from a 16-bit value.
+bool PatchIntrinsicSimplify::CanSafelyConvertTo16Bit(
+    Value* const pValue // [in] The value to check
+    ) const
+{
+    if (pValue->getType()->isHalfTy() || pValue->getType()->isIntegerTy(16))
+    {
+        // The value is already 16-bit, so we don't want to convert to 16-bit again!
+        return false;
+    }
+    else if (ConstantFP* const pConstFloat = dyn_cast<ConstantFP>(pValue))
+    {
+        // We need to check that if we cast the index down to a half, we do not lose precision.
+        APFloat apFloat(pConstFloat->getValueAPF());
+        bool losesInfo = true;
+        apFloat.convert(APFloat::IEEEhalf(), APFloat::rmTowardZero, &losesInfo);
+        return (losesInfo == false);
+    }
+    else if (CastInst* const pCast = dyn_cast<CastInst>(pValue))
```

Comment:
Can you start the checks by stripping all casts instead?

Reply:
We looked at the casts again and noticed that `CastInst` is probably not the right type to check for. For example, we do not want to accept `fptosi` from a half value to an i32. Chains of casts can also do weird stuff like sign expanding an integer, then casting to a float so we don’t want to convert those either.